tempesta_fw/tls.c

```diff
@@ -31,38 +42,169 @@
 static int
 tfw_tls_msg_process(void *conn, struct sk_buff *skb, unsigned int off)
 {
-	int r = TFW_BLOCK;
+	int r;
 	TfwConnection *c = (TfwConnection *)conn;
-	TfwMsg *msg = (TfwMsg *)c->msg;
+	TfwTlsContext *tls = c->tls;
+
+	ss_skb_queue_tail(&tls->rx_queue, skb);
+
+	r = mbedtls_ssl_handshake(&c->tls->ctx);
+	if (r == MBEDTLS_ERR_SSL_CONN_EOF) {
+		return TFW_PASS;
+	} else if (r == 0) {
+		struct sk_buff *nskb = alloc_skb(MAX_TCP_HEADER + 1024, GFP_ATOMIC);
+
+		if (unlikely(!nskb))
+			return TFW_BLOCK;
+
+		skb_reserve(nskb, MAX_TCP_HEADER);
+		skb_put(nskb, 1024);
+
+		r = mbedtls_ssl_read(&tls->ctx, nskb->data, 1024);
+
+		printk("mbedtls_ssl_read() r = %d\n", r);
+		if (r == MBEDTLS_ERR_SSL_WANT_READ || r == MBEDTLS_ERR_SSL_WANT_WRITE) {
+			kfree_skb(nskb);
+			return TFW_PASS;
+		} else if (r <= 0) {
+			kfree_skb(nskb);
+			return r ? TFW_BLOCK : TFW_PASS;
+		}
+		printk("mbedtls_ssl_read() r = %d [%.*s]\n", r, r, nskb->data);
+
+		skb_trim(nskb, r);
+
+		r = tfw_gfsm_move(&c->msg->state, TFW_HTTPS_FSM_TODO_ISSUE_81, nskb, 0);
+		if (r == TFW_BLOCK)
+			return TFW_BLOCK;
 
-	/*
-	 * TODO switch to HTTP FSM, @data and @len must be decrypted here.
-	 * Typically we don't need original skb any more either for server or
-	 * proxy modes, so it has sense to do decryption in-place.
-	 */
-	r = tfw_gfsm_move(&msg->state, TFW_HTTPS_FSM_TODO_ISSUE_81, skb, off);
+		return TFW_PASS;
+	}
 
-	return r;
+	return TFW_BLOCK;
 }
 
-/* TODO alloc TLS- or HTTPS-specific message. */
 static TfwMsg *
 tfw_tls_conn_msg_alloc(TfwConnection *conn)
 {
-	return NULL;
+	TfwHttpMsg *hm;
+
+	BUG_ON(TFW_CONN_TYPE(conn) & Conn_Srv);
+
+	hm = tfw_http_msg_alloc(TFW_CONN_TYPE(conn));
+	if (unlikely(!hm))
+		return NULL;
+
+	hm->conn = conn;
+	tfw_connection_get(conn);
+	tfw_gfsm_state_init(&hm->msg.state, conn, TFW_HTTPS_FSM_INIT);
+
+	printk("%s hm = %p\n", __func__, hm);
+
+	return (TfwMsg *)hm;
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// TODO:
+//  - use ss_skb_alloc instead of alloc_skb
+//
+
+static int
+tfw_tls_send(void *conn, const unsigned char *buf, size_t len)
+{
+	TfwConnection *c = (TfwConnection *)conn;
+	TfwTlsContext *tls = c->tls;
+	struct sk_buff *skb = alloc_skb(MAX_TCP_HEADER + len, GFP_ATOMIC);
+
+	if (unlikely(!skb))
+		return MBEDTLS_ERR_NET_SEND_FAILED;
+
+	skb_reserve(skb, MAX_TCP_HEADER);
+	skb_put(skb, len);
+
+	if (unlikely(skb_store_bits(skb, 0, buf, len)))
```

Comment:
That's just performance killer!

1. you need to allocate new skb only if it's cloned/shared, however we already handle this at SS layer;

2. `mbedtls_ssl_write()` does lots of memory copies (just see `memcpy()` and `memmove()` calls in it) to encrypt the daata and after that you make one more copy here.

Consider you call `tfw_connection_send()` just to send some static response (e.g. `tfw_http_send_502()`) - why should we allocate the skb? Tempesta code has already created skbs.